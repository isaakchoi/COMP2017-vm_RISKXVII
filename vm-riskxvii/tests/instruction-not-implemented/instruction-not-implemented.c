// Isaak Choi
// 520488399
// icho6322


/*

The following code produces a binary that includes an instruction
that hasn't been implemented. Thus, it can be used to test for
this error case.

*/



// Constants
#define READ_INT_ADDR   (0x0816)
#define WRITE_UINT_ADDR (0x0808)
#define WRITE_CHAR_ADDR (0x0800)
#define MALLOC_ADDR     (0x0830)
#define FREE_ADDR       (0x0834)
#define READ_CHAR_ADDR  (0x0812)

// Get the address of the malloced memory
int get_malloc_addr(){
    int result;
    asm volatile("addi %[res], t3, 0"
        : [res]"=r"(result));
    return result;
}

// Other functions
void print_string(char *str) {
    while (*str) *((char *)WRITE_CHAR_ADDR) = *(str++);
}

int get_int() {
    return *((int*)READ_INT_ADDR);
}

// Memory operation functions
// Not readable, but forces compiler to perform operations

void inst_sb(char b) {
    return;
}

void inst_sh(short h) {
    return;
}

void inst_sw(int w) {
    return;
}

int inst_lb() {
    int result;
    asm volatile("addi %[res], t3, 0"
        : [res]"=r"(result));
    return result;
}

int inst_lh() {
    int result;
    asm volatile("addi %[res], t3, 0"
        : [res]"=r"(result));
    return result;
}

int inst_lw() {
    int result;
    asm volatile("addi %[res], t3, 0"
        : [res]"=r"(result));
    return result;
}

int inst_lbu() {

}

int inst_lhu() {

}

int main(){

    // // Get value in form of char from user
    // print_string("Enter char to be used as byte: ");
    // char c = *(char*)READ_CHAR_ADDR;
    // print_string("\n");
    // print_string("Given char: ");
    // *(int*)WRITE_CHAR_ADDR = c;
    // print_string("\n");

    char c = 'A';

    // Constants or bitshifting - Prevent slli and srli
    int shift_n_word = 24;
    int shift_n_half = 8;

    // Store byte
    inst_sb(c);

    // Store half
    inst_sh(c << shift_n_half);

    // Store word
    inst_sw(c << shift_n_word);

    // Return
    return 0;
}

