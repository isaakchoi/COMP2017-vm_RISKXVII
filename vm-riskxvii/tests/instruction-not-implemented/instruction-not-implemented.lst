
instruction-not-implemented/instruction-not-implemented:     file format elf32-littleriscv


Disassembly of section .text:

00000000 <_start>:
   0:	7ff00113          	li	sp,2047
   4:	1cc000ef          	jal	ra,1d0 <main>
   8:	000017b7          	lui	a5,0x1
   c:	80078623          	sb	zero,-2036(a5) # 80c <ADDR_END+0xc>

00000010 <get_malloc_addr>:
  10:	fe010113          	addi	sp,sp,-32
  14:	00812e23          	sw	s0,28(sp)
  18:	02010413          	addi	s0,sp,32
  1c:	000e0793          	mv	a5,t3
  20:	fef42623          	sw	a5,-20(s0)
  24:	fec42783          	lw	a5,-20(s0)
  28:	00078513          	mv	a0,a5
  2c:	01c12403          	lw	s0,28(sp)
  30:	02010113          	addi	sp,sp,32
  34:	00008067          	ret

00000038 <print_string>:
  38:	fe010113          	addi	sp,sp,-32
  3c:	00812e23          	sw	s0,28(sp)
  40:	02010413          	addi	s0,sp,32
  44:	fea42623          	sw	a0,-20(s0)
  48:	0200006f          	j	68 <print_string+0x30>
  4c:	fec42783          	lw	a5,-20(s0)
  50:	00178713          	addi	a4,a5,1
  54:	fee42623          	sw	a4,-20(s0)
  58:	00001737          	lui	a4,0x1
  5c:	80070713          	addi	a4,a4,-2048 # 800 <ADDR_END>
  60:	0007c783          	lbu	a5,0(a5)
  64:	00f70023          	sb	a5,0(a4)
  68:	fec42783          	lw	a5,-20(s0)
  6c:	0007c783          	lbu	a5,0(a5)
  70:	fc079ee3          	bnez	a5,4c <print_string+0x14>
  74:	00000013          	nop
  78:	00000013          	nop
  7c:	01c12403          	lw	s0,28(sp)
  80:	02010113          	addi	sp,sp,32
  84:	00008067          	ret

00000088 <get_int>:
  88:	ff010113          	addi	sp,sp,-16
  8c:	00812623          	sw	s0,12(sp)
  90:	01010413          	addi	s0,sp,16
  94:	000017b7          	lui	a5,0x1
  98:	81678793          	addi	a5,a5,-2026 # 816 <ADDR_END+0x16>
  9c:	0007a783          	lw	a5,0(a5)
  a0:	00078513          	mv	a0,a5
  a4:	00c12403          	lw	s0,12(sp)
  a8:	01010113          	addi	sp,sp,16
  ac:	00008067          	ret

000000b0 <inst_sb>:
  b0:	fe010113          	addi	sp,sp,-32
  b4:	00812e23          	sw	s0,28(sp)
  b8:	02010413          	addi	s0,sp,32
  bc:	00050793          	mv	a5,a0
  c0:	fef407a3          	sb	a5,-17(s0)
  c4:	00000013          	nop
  c8:	01c12403          	lw	s0,28(sp)
  cc:	02010113          	addi	sp,sp,32
  d0:	00008067          	ret

000000d4 <inst_sh>:
  d4:	fe010113          	addi	sp,sp,-32
  d8:	00812e23          	sw	s0,28(sp)
  dc:	02010413          	addi	s0,sp,32
  e0:	00050793          	mv	a5,a0
  e4:	fef41723          	sh	a5,-18(s0)
  e8:	00000013          	nop
  ec:	01c12403          	lw	s0,28(sp)
  f0:	02010113          	addi	sp,sp,32
  f4:	00008067          	ret

000000f8 <inst_sw>:
  f8:	fe010113          	addi	sp,sp,-32
  fc:	00812e23          	sw	s0,28(sp)
 100:	02010413          	addi	s0,sp,32
 104:	fea42623          	sw	a0,-20(s0)
 108:	00000013          	nop
 10c:	01c12403          	lw	s0,28(sp)
 110:	02010113          	addi	sp,sp,32
 114:	00008067          	ret

00000118 <inst_lb>:
 118:	fe010113          	addi	sp,sp,-32
 11c:	00812e23          	sw	s0,28(sp)
 120:	02010413          	addi	s0,sp,32
 124:	000e0793          	mv	a5,t3
 128:	fef42623          	sw	a5,-20(s0)
 12c:	fec42783          	lw	a5,-20(s0)
 130:	00078513          	mv	a0,a5
 134:	01c12403          	lw	s0,28(sp)
 138:	02010113          	addi	sp,sp,32
 13c:	00008067          	ret

00000140 <inst_lh>:
 140:	fe010113          	addi	sp,sp,-32
 144:	00812e23          	sw	s0,28(sp)
 148:	02010413          	addi	s0,sp,32
 14c:	000e0793          	mv	a5,t3
 150:	fef42623          	sw	a5,-20(s0)
 154:	fec42783          	lw	a5,-20(s0)
 158:	00078513          	mv	a0,a5
 15c:	01c12403          	lw	s0,28(sp)
 160:	02010113          	addi	sp,sp,32
 164:	00008067          	ret

00000168 <inst_lw>:
 168:	fe010113          	addi	sp,sp,-32
 16c:	00812e23          	sw	s0,28(sp)
 170:	02010413          	addi	s0,sp,32
 174:	000e0793          	mv	a5,t3
 178:	fef42623          	sw	a5,-20(s0)
 17c:	fec42783          	lw	a5,-20(s0)
 180:	00078513          	mv	a0,a5
 184:	01c12403          	lw	s0,28(sp)
 188:	02010113          	addi	sp,sp,32
 18c:	00008067          	ret

00000190 <inst_lbu>:
 190:	ff010113          	addi	sp,sp,-16
 194:	00812623          	sw	s0,12(sp)
 198:	01010413          	addi	s0,sp,16
 19c:	00000013          	nop
 1a0:	00078513          	mv	a0,a5
 1a4:	00c12403          	lw	s0,12(sp)
 1a8:	01010113          	addi	sp,sp,16
 1ac:	00008067          	ret

000001b0 <inst_lhu>:
 1b0:	ff010113          	addi	sp,sp,-16
 1b4:	00812623          	sw	s0,12(sp)
 1b8:	01010413          	addi	s0,sp,16
 1bc:	00000013          	nop
 1c0:	00078513          	mv	a0,a5
 1c4:	00c12403          	lw	s0,12(sp)
 1c8:	01010113          	addi	sp,sp,16
 1cc:	00008067          	ret

000001d0 <main>:
 1d0:	fe010113          	addi	sp,sp,-32
 1d4:	00112e23          	sw	ra,28(sp)
 1d8:	00812c23          	sw	s0,24(sp)
 1dc:	02010413          	addi	s0,sp,32
 1e0:	04100793          	li	a5,65
 1e4:	fef407a3          	sb	a5,-17(s0)
 1e8:	01800793          	li	a5,24
 1ec:	fef42423          	sw	a5,-24(s0)
 1f0:	00800793          	li	a5,8
 1f4:	fef42223          	sw	a5,-28(s0)
 1f8:	fef44783          	lbu	a5,-17(s0)
 1fc:	00078513          	mv	a0,a5
 200:	eb1ff0ef          	jal	ra,b0 <inst_sb>
 204:	fef44703          	lbu	a4,-17(s0)
 208:	fe442783          	lw	a5,-28(s0)
 20c:	00f717b3          	sll	a5,a4,a5
 210:	01079793          	slli	a5,a5,0x10
 214:	4107d793          	srai	a5,a5,0x10
 218:	00078513          	mv	a0,a5
 21c:	eb9ff0ef          	jal	ra,d4 <inst_sh>
 220:	fef44703          	lbu	a4,-17(s0)
 224:	fe842783          	lw	a5,-24(s0)
 228:	00f717b3          	sll	a5,a4,a5
 22c:	00078513          	mv	a0,a5
 230:	ec9ff0ef          	jal	ra,f8 <inst_sw>
 234:	00000793          	li	a5,0
 238:	00078513          	mv	a0,a5
 23c:	01c12083          	lw	ra,28(sp)
 240:	01812403          	lw	s0,24(sp)
 244:	02010113          	addi	sp,sp,32
 248:	00008067          	ret
