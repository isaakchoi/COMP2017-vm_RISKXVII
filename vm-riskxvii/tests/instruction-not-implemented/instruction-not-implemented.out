Instruction Not Implemented: 0x01079793
PC = 0x00000210;
R[0] = 0x00000000;
R[1] = 0x00000204;
R[2] = 0x000007df;
R[3] = 0x00000000;
R[4] = 0x00000000;
R[5] = 0x00000000;
R[6] = 0x00000000;
R[7] = 0x00000000;
R[8] = 0x000007ff;
R[9] = 0x00000000;
R[10] = 0x00000041;
R[11] = 0x00000000;
R[12] = 0x00000000;
R[13] = 0x00000000;
R[14] = 0x00000041;
R[15] = 0x00004100;
R[16] = 0x00000000;
R[17] = 0x00000000;
R[18] = 0x00000000;
R[19] = 0x00000000;
R[20] = 0x00000000;
R[21] = 0x00000000;
R[22] = 0x00000000;
R[23] = 0x00000000;
R[24] = 0x00000000;
R[25] = 0x00000000;
R[26] = 0x00000000;
R[27] = 0x00000000;
R[28] = 0x00000000;
R[29] = 0x00000000;
R[30] = 0x00000000;
R[31] = 0x00000000;
